### PR TITLE
Don't answer peers with a low reputation

### PR DESCRIPTION
**Pull request by kpp**
- What does it do?

We don't have to answer to peers with low reputation for incoming requests (such as block requests).

Fixes https://github.com/paritytech/substrate/issues/8274

### CODE REVIEW

**Comment by arkpar** [comment]
Don't we already disconnect and ban peers with low reputation? How is this additional mechanism better?
I don't think deliberately ignoring requests and causing network timeouts will improve network health.

**Comment by tomaka** [comment]
> Don't we already disconnect and ban peers with low reputation? How is this additional mechanism better?

The ban only applies to opening a substream with the block announces, not to block requests. It is the problem that this PR fixes.

>  don't think deliberately ignoring requests and causing network timeouts will improve network health.

It wouldn't cause a timeout. The requesting side would be immediately notified that we have refused its request.

File: client/network/src/behaviour.rs [hunk 1]
```diff
@@ -203,7 +203,7 @@ impl<B: BlockT> Behaviour<B> {
 			discovery: disco_config.finish(),
 			bitswap: bitswap.into(),
 			request_responses:
-				request_responses::RequestResponsesBehaviour::new(request_response_protocols.into_iter())?,
+				request_responses::RequestResponsesBehaviour::new(request_response_protocols.into_iter(), None)?, // FIXME
```
Comment (kpp) [inline client/network/src/behaviour.rs:206]: @tomaka how do I get PeerHandle here?

**Comment by kpp** [comment]
Can we ban incoming connections at the networking layer, btw?

**Comment by tomaka** [comment]
> Can we ban incoming connections at the networking layer, btw?

It is possible with https://docs.rs/libp2p/0.38.0/libp2p/swarm/struct.ExpandedSwarm.html#method.ban_peer_id
However doing so would also prevent nodes from using the discovery system, which it might be too harsh.

**Comment by tomaka** [comment]
Sorry haven't taken the time to review this, but it's important.

**Comment by kpp** [comment]
ping @tomaka

**Review by tomaka — APPROVED** [review]
Sorry, this took a long time. I had a nit as to whether the reputation threshold should be made public in the peerset, but in the end I don't think it matter.

**Comment by tomaka** [comment]
bot merge

**Pull request merged**